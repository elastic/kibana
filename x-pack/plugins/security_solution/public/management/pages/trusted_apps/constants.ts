/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import type { CreateExceptionListSchema } from '@kbn/securitysolution-io-ts-list-types';
import { ExceptionListTypeEnum } from '@kbn/securitysolution-io-ts-list-types';
import {
  ENDPOINT_TRUSTED_APPS_LIST_DESCRIPTION,
  ENDPOINT_TRUSTED_APPS_LIST_ID,
  ENDPOINT_TRUSTED_APPS_LIST_NAME,
} from '@kbn/securitysolution-list-constants';

export const SEARCHABLE_FIELDS: Readonly<string[]> = [
  `name`,
  `description`,
  'item_id',
  `entries.value`,
  `entries.entries.value`,
];

export const TRUSTED_APPS_EXCEPTION_LIST_DEFINITION: CreateExceptionListSchema = {
  name: ENDPOINT_TRUSTED_APPS_LIST_NAME,
  namespace_type: 'agnostic',
  description: ENDPOINT_TRUSTED_APPS_LIST_DESCRIPTION,
  list_id: ENDPOINT_TRUSTED_APPS_LIST_ID,
  type: ExceptionListTypeEnum.ENDPOINT_TRUSTED_APPS,
};
