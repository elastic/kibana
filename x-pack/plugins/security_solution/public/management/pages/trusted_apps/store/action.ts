/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { Action } from 'redux';

import { NewTrustedApp, TrustedApp } from '../../../../../common/endpoint/types';
import { AsyncResourceState, TrustedAppsListData } from '../state';
import { GetPolicyListResponse } from '../../policy/types';

export type TrustedAppsListDataOutdated = Action<'trustedAppsListDataOutdated'>;

interface ResourceStateChanged<T, D = null> extends Action<T> {
  payload: { newState: AsyncResourceState<D> };
}

export type TrustedAppsListResourceStateChanged = ResourceStateChanged<
  'trustedAppsListResourceStateChanged',
  TrustedAppsListData
>;

export type TrustedAppDeletionSubmissionResourceStateChanged = ResourceStateChanged<'trustedAppDeletionSubmissionResourceStateChanged'>;

export type TrustedAppDeletionDialogStarted = Action<'trustedAppDeletionDialogStarted'> & {
  payload: {
    entry: TrustedApp;
  };
};

export type TrustedAppDeletionDialogConfirmed = Action<'trustedAppDeletionDialogConfirmed'>;

export type TrustedAppDeletionDialogClosed = Action<'trustedAppDeletionDialogClosed'>;

export type TrustedAppCreationSubmissionResourceStateChanged = ResourceStateChanged<
  'trustedAppCreationSubmissionResourceStateChanged',
  TrustedApp
>;

export type TrustedAppCreationDialogStarted = Action<'trustedAppCreationDialogStarted'> & {
  payload: {
    entry: NewTrustedApp;
  };
};

export type TrustedAppCreationDialogFormStateUpdated = Action<'trustedAppCreationDialogFormStateUpdated'> & {
  payload: {
    entry: NewTrustedApp;
    isValid: boolean;
  };
};

export type TrustedAppCreationEditItemStateChanged = Action<'trustedAppCreationEditItemStateChanged'> & {
  payload: AsyncResourceState<TrustedApp>;
};

export type TrustedAppCreationDialogConfirmed = Action<'trustedAppCreationDialogConfirmed'>;

export type TrustedAppCreationDialogClosed = Action<'trustedAppCreationDialogClosed'>;

export type TrustedAppsExistResponse = Action<'trustedAppsExistStateChanged'> & {
  payload: AsyncResourceState<boolean>;
};

export type TrustedAppsPoliciesStateChanged = Action<'trustedAppsPoliciesStateChanged'> & {
  payload: AsyncResourceState<GetPolicyListResponse>;
};

export type TrustedAppForceRefresh = Action<'trustedAppForceRefresh'> & {
  payload: {
    forceRefresh: boolean;
  };
};

export type TrustedAppsPageAction =
  | TrustedAppsListDataOutdated
  | TrustedAppsListResourceStateChanged
  | TrustedAppDeletionSubmissionResourceStateChanged
  | TrustedAppDeletionDialogStarted
  | TrustedAppDeletionDialogConfirmed
  | TrustedAppDeletionDialogClosed
  | TrustedAppCreationSubmissionResourceStateChanged
  | TrustedAppCreationEditItemStateChanged
  | TrustedAppCreationDialogStarted
  | TrustedAppCreationDialogFormStateUpdated
  | TrustedAppCreationDialogConfirmed
  | TrustedAppsExistResponse
  | TrustedAppsPoliciesStateChanged
  | TrustedAppCreationDialogClosed
  | TrustedAppForceRefresh;
