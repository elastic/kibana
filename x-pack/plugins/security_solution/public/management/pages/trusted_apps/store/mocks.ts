/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { GetPolicyListResponse } from '../../policy/types';

import { EndpointDocGenerator } from '../../../../../common/endpoint/generate_data';

export const getGeneratedPolicyResponse = (): GetPolicyListResponse => ({
  items: [new EndpointDocGenerator('seed').generatePolicyPackagePolicy()],
  total: 1,
  perPage: 1,
  page: 1,
});
