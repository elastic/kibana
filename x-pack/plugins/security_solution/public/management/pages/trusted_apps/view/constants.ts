/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { i18n } from '@kbn/i18n';
import { TrustedApp } from '../../../../../common/endpoint/types';

export const OS_TITLES: Readonly<{ [K in TrustedApp['os']]: string }> = {
  windows: i18n.translate('xpack.securitySolution.trustedapps.os.windows', {
    defaultMessage: 'Windows',
  }),
  macos: i18n.translate('xpack.securitySolution.trustedapps.os.macos', {
    defaultMessage: 'Mac OS',
  }),
  linux: i18n.translate('xpack.securitySolution.trustedapps.os.linux', {
    defaultMessage: 'Linux',
  }),
};
