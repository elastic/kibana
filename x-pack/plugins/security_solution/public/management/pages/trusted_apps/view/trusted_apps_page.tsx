/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React, { memo, useMemo } from 'react';
import { useLocation } from 'react-router-dom';
import styled from 'styled-components';
import { FormattedMessage } from '@kbn/i18n/react';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiEmptyPrompt,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiLoadingSpinner,
  EuiSpacer,
} from '@elastic/eui';

import { ViewType } from '../state';
import {
  checkingIfEntriesExist,
  entriesExist,
  getCurrentLocation,
  getListTotalItemsCount,
} from '../store/selectors';
import { useTrustedAppsNavigateCallback, useTrustedAppsSelector } from './hooks';
import { AdministrationListPage } from '../../../components/administration_list_page';
import { CreateTrustedAppFlyout } from './components/create_trusted_app_flyout';
import { ControlPanel } from './components/control_panel';
import { TrustedAppsGrid } from './components/trusted_apps_grid';
import { TrustedAppsList } from './components/trusted_apps_list';
import { TrustedAppDeletionDialog } from './trusted_app_deletion_dialog';
import { TrustedAppsNotifications } from './trusted_apps_notifications';
import { TrustedAppsListPageRouteState } from '../../../../../common/endpoint/types';
import { useNavigateToAppEventHandler } from '../../../../common/hooks/endpoint/use_navigate_to_app_event_handler';
import { ABOUT_TRUSTED_APPS } from './translations';
import { EmptyState } from './components/empty_state';
import { SearchBar } from '../../../components/search_bar';

export const TrustedAppsPage = memo(() => {
  const { state: routeState } = useLocation<TrustedAppsListPageRouteState | undefined>();
  const location = useTrustedAppsSelector(getCurrentLocation);
  const totalItemsCount = useTrustedAppsSelector(getListTotalItemsCount);
  const isCheckingIfEntriesExists = useTrustedAppsSelector(checkingIfEntriesExist);
  const doEntriesExist = useTrustedAppsSelector(entriesExist) === true;
  const handleAddButtonClick = useTrustedAppsNavigateCallback(() => ({
    show: 'create',
    id: undefined,
  }));
  const handleAddFlyoutClose = useTrustedAppsNavigateCallback(() => ({
    show: undefined,
    id: undefined,
  }));
  const handleViewTypeChange = useTrustedAppsNavigateCallback((viewType: ViewType) => ({
    view_type: viewType,
  }));
  const handleOnSearch = useTrustedAppsNavigateCallback((query: string) => ({ filter: query }));

  const showCreateFlyout = !!location.show;

  const backButton = useMemo(() => {
    if (routeState && routeState.onBackButtonNavigateTo) {
      return <BackToExternalAppButton {...routeState} />;
    }
    return null;
  }, [routeState]);

  const addButton = (
    <EuiButton
      fill
      iconType="plusInCircle"
      isDisabled={showCreateFlyout}
      onClick={handleAddButtonClick}
      data-test-subj="trustedAppsListAddButton"
    >
      <FormattedMessage
        id="xpack.securitySolution.trustedapps.list.addButton"
        defaultMessage="Add Trusted Application"
      />
    </EuiButton>
  );

  const content = (
    <>
      <TrustedAppDeletionDialog />

      {showCreateFlyout && (
        <CreateTrustedAppFlyout
          onClose={handleAddFlyoutClose}
          size="m"
          data-test-subj="addTrustedAppFlyout"
        />
      )}

      <SearchBar defaultValue={location.filter} onSearch={handleOnSearch} />
      {doEntriesExist ? (
        <EuiFlexGroup
          direction="column"
          gutterSize="none"
          data-test-subj="trustedAppsListPageContent"
        >
          <EuiSpacer size="m" />
          <EuiFlexItem grow={false}>
            <ControlPanel
              totalItemCount={totalItemsCount}
              currentViewType={location.view_type}
              onViewTypeChange={handleViewTypeChange}
            />

            <EuiSpacer size="m" />
          </EuiFlexItem>
          <EuiFlexItem>
            <EuiHorizontalRule margin="none" />

            {location.view_type === 'grid' && <TrustedAppsGrid />}
            {location.view_type === 'list' && <TrustedAppsList />}
          </EuiFlexItem>
        </EuiFlexGroup>
      ) : (
        <EmptyState onAdd={handleAddButtonClick} isAddDisabled={showCreateFlyout} />
      )}
    </>
  );

  return (
    <AdministrationListPage
      data-test-subj="trustedAppsListPage"
      beta={false}
      title={
        <FormattedMessage
          id="xpack.securitySolution.trustedapps.list.pageTitle"
          defaultMessage="Trusted Applications"
        />
      }
      headerBackComponent={backButton}
      subtitle={ABOUT_TRUSTED_APPS}
      actions={doEntriesExist ? addButton : <></>}
    >
      <TrustedAppsNotifications />

      {isCheckingIfEntriesExists ? (
        <EuiEmptyPrompt
          data-test-subj="trustedAppsListLoader"
          body={<EuiLoadingSpinner className="essentialAnimation" size="xl" />}
        />
      ) : (
        content
      )}
    </AdministrationListPage>
  );
});

TrustedAppsPage.displayName = 'TrustedAppsPage';

const EuiButtonEmptyStyled = styled(EuiButtonEmpty)`
  margin-bottom: ${({ theme }) => theme.eui.euiSizeS};

  .euiIcon {
    width: ${({ theme }) => theme.eui.euiIconSizes.small};
    height: ${({ theme }) => theme.eui.euiIconSizes.small};
  }

  .text {
    font-size: ${({ theme }) => theme.eui.euiFontSizeXS};
  }
`;

const BackToExternalAppButton = memo<TrustedAppsListPageRouteState>(
  ({ backButtonLabel, backButtonUrl, onBackButtonNavigateTo }) => {
    const handleBackOnClick = useNavigateToAppEventHandler(...onBackButtonNavigateTo!);

    return (
      <EuiButtonEmptyStyled
        flush="left"
        size="xs"
        iconType="arrowLeft"
        href={backButtonUrl!}
        onClick={handleBackOnClick}
        textProps={{ className: 'text' }}
        data-test-subj="backToOrigin"
      >
        {backButtonLabel || (
          <FormattedMessage
            id="xpack.securitySolution.trustedapps.list.backButton"
            defaultMessage="Back"
          />
        )}
      </EuiButtonEmptyStyled>
    );
  }
);

BackToExternalAppButton.displayName = 'BackToExternalAppButton';
