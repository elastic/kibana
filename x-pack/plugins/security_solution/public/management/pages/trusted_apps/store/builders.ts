/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import {
  ConditionEntry,
  ConditionEntryField,
  NewTrustedApp,
  OperatingSystem,
} from '../../../../../common/endpoint/types';

import { MANAGEMENT_DEFAULT_PAGE, MANAGEMENT_DEFAULT_PAGE_SIZE } from '../../../common/constants';

import { TrustedAppsListPageState } from '../state';

export const defaultConditionEntry = (): ConditionEntry<ConditionEntryField.HASH> => ({
  field: ConditionEntryField.HASH,
  operator: 'included',
  type: 'match',
  value: '',
});

export const defaultNewTrustedApp = (): NewTrustedApp => ({
  name: '',
  os: OperatingSystem.WINDOWS,
  entries: [defaultConditionEntry()],
  description: '',
});

export const initialDeletionDialogState = (): TrustedAppsListPageState['deletionDialog'] => ({
  confirmed: false,
  submissionResourceState: { type: 'UninitialisedResourceState' },
});

export const initialCreationDialogState = (): TrustedAppsListPageState['creationDialog'] => ({
  confirmed: false,
  submissionResourceState: { type: 'UninitialisedResourceState' },
});

export const initialTrustedAppsPageState = (): TrustedAppsListPageState => ({
  entriesExist: { type: 'UninitialisedResourceState' },
  listView: {
    listResourceState: { type: 'UninitialisedResourceState' },
    freshDataTimestamp: Date.now(),
  },
  deletionDialog: initialDeletionDialogState(),
  creationDialog: initialCreationDialogState(),
  location: {
    page_index: MANAGEMENT_DEFAULT_PAGE,
    page_size: MANAGEMENT_DEFAULT_PAGE_SIZE,
    show: undefined,
    view_type: 'grid',
  },
  active: false,
});
