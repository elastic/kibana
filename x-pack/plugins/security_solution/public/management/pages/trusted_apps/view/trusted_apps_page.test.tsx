/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React from 'react';
import * as reactTestingLibrary from '@testing-library/react';
import { TrustedAppsPage } from './trusted_apps_page';
import { AppContextTestRender, createAppRootMockRenderer } from '../../../../common/mock/endpoint';
import { fireEvent } from '@testing-library/dom';
import { MiddlewareActionSpyHelper } from '../../../../common/store/test_utils';
import {
  ConditionEntryField,
  GetTrustedListAppsResponse,
  NewTrustedApp,
  OperatingSystem,
  PostTrustedAppCreateResponse,
  TrustedApp,
} from '../../../../../common/endpoint/types';
import { HttpFetchOptions } from 'kibana/public';
import { TRUSTED_APPS_LIST_API } from '../../../../../common/endpoint/constants';

jest.mock('@elastic/eui/lib/services/accessibility/html_id_generator', () => ({
  htmlIdGenerator: () => () => 'mockId',
}));

describe('When on the Trusted Apps Page', () => {
  const expectedAboutInfo =
    'Add a trusted application to improve performance or alleviate conflicts with other applications running on your hosts. Trusted applications will be applied to hosts running Endpoint Security.';

  let mockedContext: AppContextTestRender;
  let history: AppContextTestRender['history'];
  let coreStart: AppContextTestRender['coreStart'];
  let waitForAction: MiddlewareActionSpyHelper['waitForAction'];
  let render: () => ReturnType<AppContextTestRender['render']>;
  const originalScrollTo = window.scrollTo;
  const act = reactTestingLibrary.act;

  const getFakeTrustedApp = (): TrustedApp => ({
    id: '1111-2222-3333-4444',
    name: 'one app',
    os: OperatingSystem.WINDOWS,
    created_at: '2021-01-04T13:55:00.561Z',
    created_by: 'me',
    description: 'a good one',
    entries: [
      {
        field: ConditionEntryField.PATH,
        value: 'one/two',
        operator: 'included',
        type: 'match',
      },
    ],
  });

  const mockListApis = (http: AppContextTestRender['coreStart']['http']) => {
    const currentGetHandler = http.get.getMockImplementation();

    http.get.mockImplementation(async (...args) => {
      const path = (args[0] as unknown) as string;
      // @ts-ignore
      const httpOptions = args[1] as HttpFetchOptions;

      if (path === TRUSTED_APPS_LIST_API) {
        return {
          data: [getFakeTrustedApp()],
          total: 50, // << Should be a value large enough to fulfill two pages
          page: httpOptions?.query?.page ?? 1,
          per_page: httpOptions?.query?.per_page ?? 20,
        };
      }
      if (currentGetHandler) {
        return currentGetHandler(...args);
      }
    });
  };

  beforeAll(() => {
    window.scrollTo = () => {};
  });

  afterAll(() => {
    window.scrollTo = originalScrollTo;
  });

  beforeEach(() => {
    mockedContext = createAppRootMockRenderer();

    history = mockedContext.history;
    coreStart = mockedContext.coreStart;
    waitForAction = mockedContext.middlewareSpy.waitForAction;
    render = () => mockedContext.render(<TrustedAppsPage />);
    reactTestingLibrary.act(() => {
      history.push('/trusted_apps');
    });
    window.scrollTo = jest.fn();
  });

  describe('and there is trusted app entries', () => {
    const renderWithListData = async () => {
      const renderResult = render();
      await act(async () => {
        await waitForAction('trustedAppsListResourceStateChanged');
      });
      return renderResult;
    };

    beforeEach(() => mockListApis(coreStart.http));

    it('should display subtitle info about trusted apps', async () => {
      const { getByTestId } = await renderWithListData();
      expect(getByTestId('header-panel-subtitle').textContent).toEqual(expectedAboutInfo);
    });

    it('should display a Add Trusted App button', async () => {
      const { getByTestId } = await renderWithListData();
      const addButton = await getByTestId('trustedAppsListAddButton');
      expect(addButton.textContent).toBe('Add Trusted Application');
    });
  });

  describe('when the Add Trusted App button is clicked', () => {
    const renderAndClickAddButton = async (): Promise<
      ReturnType<AppContextTestRender['render']>
    > => {
      const renderResult = render();
      await act(async () => {
        await waitForAction('trustedAppsListResourceStateChanged');
      });
      const addButton = renderResult.getByTestId('trustedAppsListAddButton');
      reactTestingLibrary.act(() => {
        fireEvent.click(addButton, { button: 1 });
      });
      return renderResult;
    };

    beforeEach(() => mockListApis(coreStart.http));

    it('should display the create flyout', async () => {
      const { getByTestId } = await renderAndClickAddButton();
      const flyout = getByTestId('addTrustedAppFlyout');
      expect(flyout).not.toBeNull();

      const flyoutTitle = getByTestId('addTrustedAppFlyout-headerTitle');
      expect(flyoutTitle.textContent).toBe('Add trusted application');
    });

    it('should update the URL to indicate the flyout is opened', async () => {
      await renderAndClickAddButton();
      expect(/show\=create/.test(history.location.search)).toBe(true);
    });

    it('should preserve other URL search params', async () => {
      reactTestingLibrary.act(() => {
        history.push('/trusted_apps?page_index=2&page_size=20');
      });
      await renderAndClickAddButton();
      expect(history.location.search).toBe('?page_index=2&page_size=20&show=create');
    });

    it('should display create form', async () => {
      const { queryByTestId } = await renderAndClickAddButton();
      expect(queryByTestId('addTrustedAppFlyout-createForm')).not.toBeNull();
    });

    it('should initially have the flyout Add button disabled', async () => {
      const { getByTestId } = await renderAndClickAddButton();
      expect((getByTestId('addTrustedAppFlyout-createButton') as HTMLButtonElement).disabled).toBe(
        true
      );
    });

    it('should close flyout if cancel button is clicked', async () => {
      const { getByTestId, queryByTestId } = await renderAndClickAddButton();
      const cancelButton = getByTestId('addTrustedAppFlyout-cancelButton');
      reactTestingLibrary.act(() => {
        fireEvent.click(cancelButton, { button: 1 });
      });
      expect(queryByTestId('addTrustedAppFlyout')).toBeNull();
      expect(history.location.search).toBe('');
    });

    it('should close flyout if flyout close button is clicked', async () => {
      const { getByTestId, queryByTestId } = await renderAndClickAddButton();
      const flyoutCloseButton = getByTestId('euiFlyoutCloseButton');
      reactTestingLibrary.act(() => {
        fireEvent.click(flyoutCloseButton, { button: 1 });
      });
      expect(queryByTestId('addTrustedAppFlyout')).toBeNull();
      expect(history.location.search).toBe('');
    });

    describe('and when the form data is valid', () => {
      const fillInCreateForm = ({ getByTestId }: ReturnType<AppContextTestRender['render']>) => {
        reactTestingLibrary.act(() => {
          fireEvent.change(getByTestId('addTrustedAppFlyout-createForm-nameTextField'), {
            target: { value: 'trusted app A' },
          });

          fireEvent.change(
            getByTestId('addTrustedAppFlyout-createForm-conditionsBuilder-group1-entry0-value'),
            { target: { value: '44ed10b389dbcd1cf16cec79d16d7378' } }
          );

          fireEvent.change(getByTestId('addTrustedAppFlyout-createForm-descriptionField'), {
            target: { value: 'let this be' },
          });
        });
      };

      it('should enable the Flyout Add button', async () => {
        const renderResult = await renderAndClickAddButton();
        const { getByTestId } = renderResult;
        fillInCreateForm(renderResult);
        const flyoutAddButton = getByTestId(
          'addTrustedAppFlyout-createButton'
        ) as HTMLButtonElement;
        expect(flyoutAddButton.disabled).toBe(false);
      });

      describe('and the Flyout Add button is clicked', () => {
        let renderResult: ReturnType<AppContextTestRender['render']>;
        let resolveHttpPost: (response?: PostTrustedAppCreateResponse) => void;
        let httpPostBody: string;
        let rejectHttpPost: (response: Error) => void;

        beforeEach(async () => {
          // Mock the http.post() call and expose `resolveHttpPost()` method so that
          // we can control when the API call response is returned, which will allow us
          // to test the UI behaviours while the API call is in flight
          coreStart.http.post.mockImplementation(
            // @ts-ignore
            async (path: string, options: HttpFetchOptions) => {
              return new Promise((resolve, reject) => {
                httpPostBody = options.body as string;
                resolveHttpPost = resolve;
                rejectHttpPost = reject;
              });
            }
          );

          renderResult = await renderAndClickAddButton();
          fillInCreateForm(renderResult);
          const userClickedSaveActionWatcher = waitForAction('trustedAppCreationDialogConfirmed');
          reactTestingLibrary.act(() => {
            fireEvent.click(renderResult.getByTestId('addTrustedAppFlyout-createButton'), {
              button: 1,
            });
          });
          await reactTestingLibrary.act(async () => {
            await userClickedSaveActionWatcher;
          });
        });

        afterEach(() => resolveHttpPost());

        it('should display info about Trusted Apps', async () => {
          expect(renderResult.getByTestId('addTrustedAppFlyout-about').textContent).toEqual(
            expectedAboutInfo
          );
        });

        it('should disable the Cancel button', async () => {
          expect(
            (renderResult.getByTestId('addTrustedAppFlyout-cancelButton') as HTMLButtonElement)
              .disabled
          ).toBe(true);
          resolveHttpPost();
        });

        it('should hide the dialog close button', async () => {
          expect(renderResult.queryByTestId('euiFlyoutCloseButton')).toBeNull();
        });

        it('should disable the flyout Add button and set it to loading', async () => {
          const saveButton = renderResult.getByTestId(
            'addTrustedAppFlyout-createButton'
          ) as HTMLButtonElement;
          expect(saveButton.disabled).toBe(true);
          expect(saveButton.querySelector('.euiLoadingSpinner')).not.toBeNull();
        });

        describe('and if create was successful', () => {
          beforeEach(async () => {
            const successCreateApiResponse: PostTrustedAppCreateResponse = {
              data: {
                ...(JSON.parse(httpPostBody) as NewTrustedApp),
                id: '1',
                created_at: '2020-09-16T14:09:45.484Z',
                created_by: 'kibana',
              },
            };
            await reactTestingLibrary.act(async () => {
              const serverResponseAction = waitForAction(
                'trustedAppCreationSubmissionResourceStateChanged'
              );
              coreStart.http.get.mockClear();
              resolveHttpPost(successCreateApiResponse);
              await serverResponseAction;
            });
          });

          it('should close the flyout', async () => {
            expect(renderResult.queryByTestId('addTrustedAppFlyout')).toBeNull();
          });

          it('should show success toast notification', async () => {
            expect(coreStart.notifications.toasts.addSuccess.mock.calls[0][0]).toEqual(
              '"trusted app A" has been added to the Trusted Applications list.'
            );
          });

          it('should trigger the List to reload', async () => {
            expect(coreStart.http.get.mock.calls[0][0]).toEqual(TRUSTED_APPS_LIST_API);
          });
        });

        describe('and if create failed', () => {
          beforeEach(async () => {
            const failedCreateApiResponse: Error & { body?: { message: string } } = new Error(
              'Bad call'
            );
            failedCreateApiResponse.body = {
              message: 'bad call',
            };
            await reactTestingLibrary.act(async () => {
              const serverResponseAction = waitForAction(
                'trustedAppCreationSubmissionResourceStateChanged'
              );
              coreStart.http.get.mockClear();
              rejectHttpPost(failedCreateApiResponse);
              await serverResponseAction;
            });
          });

          it('should continue to show the flyout', async () => {
            expect(renderResult.getByTestId('addTrustedAppFlyout')).not.toBeNull();
          });

          it('should enable the Cancel Button', async () => {
            expect(
              (renderResult.getByTestId('addTrustedAppFlyout-cancelButton') as HTMLButtonElement)
                .disabled
            ).toBe(false);
          });

          it('should show the dialog close button', async () => {
            expect(renderResult.getByTestId('euiFlyoutCloseButton')).not.toBeNull();
          });

          it('should enable the flyout Add button and remove loading indicating', () => {
            expect(
              (renderResult.getByTestId('addTrustedAppFlyout-createButton') as HTMLButtonElement)
                .disabled
            ).toBe(false);
          });

          it('should show API errors in the form', async () => {
            expect(renderResult.container.querySelector('.euiForm__errors')).not.toBeNull();
          });
        });
      });
    });

    describe('and when the form data is not valid', () => {
      it('should not enable the Flyout Add button with an invalid hash', async () => {
        const renderResult = await renderAndClickAddButton();
        const { getByTestId } = renderResult;

        reactTestingLibrary.act(() => {
          fireEvent.change(getByTestId('addTrustedAppFlyout-createForm-nameTextField'), {
            target: { value: 'trusted app A' },
          });

          fireEvent.change(
            getByTestId('addTrustedAppFlyout-createForm-conditionsBuilder-group1-entry0-value'),
            { target: { value: 'invalid hash' } }
          );
        });

        const flyoutAddButton = getByTestId(
          'addTrustedAppFlyout-createButton'
        ) as HTMLButtonElement;
        expect(flyoutAddButton.disabled).toBe(true);
      });
    });
  });

  describe('and there are no trusted apps', () => {
    const releaseExistsResponse: jest.MockedFunction<
      () => Promise<GetTrustedListAppsResponse>
    > = jest.fn(async () => {
      return {
        data: [],
        total: 0,
        page: 1,
        per_page: 1,
      };
    });
    const releaseListResponse: jest.MockedFunction<
      () => Promise<GetTrustedListAppsResponse>
    > = jest.fn(async () => {
      return {
        data: [],
        total: 0,
        page: 1,
        per_page: 20,
      };
    });

    beforeEach(() => {
      // @ts-ignore
      coreStart.http.get.mockImplementation(async (path, options) => {
        if (path === TRUSTED_APPS_LIST_API) {
          const { page, per_page: perPage } = options.query as { page: number; per_page: number };

          if (page === 1 && perPage === 1) {
            return releaseExistsResponse();
          } else {
            return releaseListResponse();
          }
        }
      });
    });

    afterEach(() => {
      releaseExistsResponse.mockClear();
      releaseListResponse.mockClear();
    });

    it('should show a loader until trusted apps existence can be confirmed', async () => {
      // Make the call that checks if Trusted Apps exists not respond back
      releaseExistsResponse.mockImplementationOnce(() => new Promise(() => {}));
      const renderResult = render();
      expect(await renderResult.findByTestId('trustedAppsListLoader')).not.toBeNull();
    });

    it('should show Empty Prompt if not entries exist', async () => {
      const renderResult = render();
      await act(async () => {
        await waitForAction('trustedAppsExistStateChanged');
      });
      expect(await renderResult.findByTestId('trustedAppEmptyState')).not.toBeNull();
    });

    it('should hide empty prompt and show list after one trusted app is added', async () => {
      const renderResult = render();
      await act(async () => {
        await waitForAction('trustedAppsExistStateChanged');
      });
      expect(await renderResult.findByTestId('trustedAppEmptyState')).not.toBeNull();
      releaseListResponse.mockResolvedValueOnce({
        data: [getFakeTrustedApp()],
        total: 1,
        page: 1,
        per_page: 20,
      });
      releaseExistsResponse.mockResolvedValueOnce({
        data: [getFakeTrustedApp()],
        total: 1,
        page: 1,
        per_page: 1,
      });

      await act(async () => {
        mockedContext.store.dispatch({
          type: 'trustedAppsListDataOutdated',
        });
        await waitForAction('trustedAppsListResourceStateChanged');
      });

      expect(await renderResult.findByTestId('trustedAppsListPageContent')).not.toBeNull();
    });

    it('should should show empty prompt once the last trusted app entry is deleted', async () => {
      releaseListResponse.mockResolvedValueOnce({
        data: [getFakeTrustedApp()],
        total: 1,
        page: 1,
        per_page: 20,
      });
      releaseExistsResponse.mockResolvedValueOnce({
        data: [getFakeTrustedApp()],
        total: 1,
        page: 1,
        per_page: 1,
      });

      const renderResult = render();

      await act(async () => {
        await waitForAction('trustedAppsExistStateChanged');
      });

      expect(await renderResult.findByTestId('trustedAppsListPageContent')).not.toBeNull();

      releaseListResponse.mockResolvedValueOnce({
        data: [],
        total: 0,
        page: 1,
        per_page: 20,
      });
      releaseExistsResponse.mockResolvedValueOnce({
        data: [],
        total: 0,
        page: 1,
        per_page: 1,
      });

      await act(async () => {
        mockedContext.store.dispatch({
          type: 'trustedAppsListDataOutdated',
        });
        await waitForAction('trustedAppsListResourceStateChanged');
      });

      expect(await renderResult.findByTestId('trustedAppEmptyState')).not.toBeNull();
    });
  });
});
