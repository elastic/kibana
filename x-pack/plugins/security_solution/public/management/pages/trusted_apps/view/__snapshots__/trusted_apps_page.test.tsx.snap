// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrustedAppsPage rendering 1`] = `
<AdministrationListPage
  beta={true}
  subtitle={
    <FormattedMessage
      defaultMessage="View and configure trusted applications"
      id="xpack.securitySolution.trustedapps.list.pageSubTitle"
      values={Object {}}
    />
  }
  title={
    <FormattedMessage
      defaultMessage="Trusted Applications"
      id="xpack.securitySolution.trustedapps.list.pageTitle"
      values={Object {}}
    />
  }
/>
`;
